system/jlib/jstreamhelpers.hpp

```diff
@@ -15,27 +15,103 @@
     limitations under the License.
 ############################################################################## */
 
-
-
 #ifndef JSTREAMHELPERS_HPP
 #define JSTREAMHELPERS_HPP
 
 #include "platform.h"
 #include "jstream.hpp"
 
 // Global helper functions for buffered serial output
-inline void append(IBufferedSerialOutputStream & target, const char * value)
+inline void append(IBufferedSerialOutputStream &target, const char *value)
 {
     if (value)
-        target.put(strlen(value)+1, value);
+        target.put(strlen(value) + 1, value);
     else
         target.put(1, "");
 }
 
 template <class T>
-inline void append(IBufferedSerialOutputStream & target, const T & value)
+inline void append(IBufferedSerialOutputStream &target, const T &value)
 {
     target.put(sizeof(T), &value);
 }
 
+// Global helper functions for buffered serial input
+inline void read(IBufferedSerialInputStream &source, char *buffer, size32_t size)
+{
+    source.read(size, buffer);
+}
+
+template <class T>
+inline void read(IBufferedSerialInputStream &source, T &value)
+{
+    source.read(sizeof(T), &value);
+}
+
+const byte *readDirect(IBufferedSerialInputStream &source, size32_t len)
+{
+    if (len)
+    {
+        size32_t got;
+        const byte *result = static_cast<const byte *>(source.peek(len, got));
+        if (got >= len)
+        {
+            source.skip(len);
+            return result;
```

Comment:
This is not currently valid (although it probably works at the moment).  As soon as skip() is called, there is no guarantee that the peeked value is valid.  Once multithreaded input is implemented I can see cases where this would break.